### DISCUSSION TITLE
New tree and forest generative algorithms

### DISCUSSION THREAD

**Discussion by vigna** [networkx/networkx#6506] [Show and tell]
We have implemented a number of generative procedures for:

- unlabelled unrooted trees
- unlabelled trees
- unlabelled rooted forests (whose trees can be limited in size)

Presently we could find in networkx only generative procedures for labeled trees based on Prüfer sequences. The techniques for generating uniformly the unlabelled structures (i.e., up to isomorphism) above are a bit convoluted but uniform generation up to isomorphism is often what one would want to do.

Is networkx interested in a PR with these generative methods?

**Comment by dschult** [comment]
If I understand you correctly, you know of algorithms for random graph generators that select uniformly among all unlabeled tress (up to isomorphism). Yes -- we would be interested in a PR with these functions.  The trees.py module in `networkx/generators` seems like a logical place to put it, but if it is easier to make a new module that's fine too.

**Comment by vigna** [comment]
Yes. It's [Wilf's algorithm](https://www.sciencedirect.com/science/article/abs/pii/0196677481900213). To implement random generation of unrooted unlabelled trees, you need rooted unlabelled trees and forests.

Just to avoid to much fiddling back-and-forth, which names should we use? `random_tree()` at this point is taken for labelled trees. We would have:
- `random_unlabelled_tree()`
- `random_unlabelled_rooted_tree()`
- `random_unlabelled_rooted_forest()`

For all these, we have counting functions that are necessary for the generation. Do you want them exposed? That could be
- `num_random_unlabelled_tree()`
- `num_random_unlabelled_rooted_tree()`
- `num_random_unlabelled_rooted_forest()`

**Reply by dschult** [reply]
How long are these `num_random...` functions? If they essentially just compute a formula for the number of non-isomorphic trees then we could put the formula in the public function doc_strings and have these functions be private helper functions (i.e. name starting an underscore). If they are more involved, they can be public functions. I think that decision can be made once the PR is created so we can see the tradeoffs more easily. So, make a version with whatever is easiest/best for you.

As for the names for the public function, would emphasizing the uniform nature be advantageous: `uniformly_random_tree`? Or is `unlabelled` more clear?  I tend to avoid words that are spelled differently in different parts of the world (like `travelling_wave` vs `traveling_wave`) and `unlabelled` falls in that category (unlabeled). But that's probably a silly thing to base the choice of name on. Other thoughts on names?

**Comment by vigna** [comment]
No, they're not long, e.g., for the number of rooted trees:
```
    def n_ranrut(n):
        if n == 1:
            return 1
        else:
            return sum([sum([(d * n_ranrut(n - j * d)) * n_ranrut(d) for j in range(1, (n - 1) // d + 1)]) for d in range(1, n)]) // (n - 1)
```
"Uniformly" is not correct because also `random_tree()` performs uniform generation—only, on a different space. The standard term is "unlabelled", or "up to isomorphism", but we thought that `random_tree_up_to_isomorphism()` as a bit heavy. What about `random_unlbd_tree()` to avoid the problem? I think everybody would ready their own version of "unlabel[l]ed" in there...

**Reply by dschult** [reply]
OK...  I looked at these function again.  I think we should combine the current `random_tree` with the new one and add a keyword argument for `labeled=True`.  The code for the labeled version is quite short so should fit into another function easily.
That doesn't address the question of the rooted tree functions. But I think we could just make the three functions:
`random_tree`     with a `labeled=True` keyword default for backward compatibility.
`random_rooted_tree` no option for labeled => only unlabeled
`random_rooted_forest` no option for labeled => only unlabeled

The choice of spelling the keyword `labeled` matches the spelling networkx has used in other places such as `dfs_labeled_edges`.

And aren't the unlabeled_rooted_trees also called polya_trees?  Would that be a better term for a function name? And are there such a thing as polya_forests?  :)

**Comment by vigna** [comment]
Seems very reasonable.

Yes, some people in combinatorics uses "Polya trees" for unlabelled rooted trees but I think it would be awkward for most people. I think adding the keyword `labeled` is the way to go.
As for random_rooted_tree, I think that generation of random rooted labeled tree is trivial, because you generate the tree and then the root. Since there are no isomorphism, this will give you a generation procedure. I don't know if we want to add that option.

Another issue is what to return for rooted structures. Presently we always generate random trees with root the vertex 0, but thinking of the future and the possibility of generating random labeled rooted trees maybe it's better to return a tree and an integer, the root (which in our case would be always zero, presently). In any case, for forests we return a pair—an array of trees and an array of roots. Also in this case, the roots can be derived algorithmically (the first tree has root 0, the second tree has root equal to the size of the first tree, and so on) but it seems much more practical to provide the roots directly. For uniformity, we could do the same for trees.

**Comment by mjschwenne** [comment]
There is also `random_spanning_tree` which is currently located in `algorithms/tree/mst.py`. This algorithm returns a random tree that uses the edge weights as a multiplicative weights (algorithm by V.G Kulkarni [here](https://www.sciencedirect.com/science/article/abs/pii/019667749090002V), I could not fine a PDF online without a pay wall but found it at my university library) that may be relevant to this discussion. 

At the very least we may want to consolidate all of these functions into one location.

**Comment by vigna** [comment]
There's another important design decision. All these procedures perform significant recursion, and we are implementing everything with explicit stacks to circumvent Python's bizarre limits. But they really benefit from some temporary caching. If you call them more than once, the cache can actually be reused. LRU cache decorators don't solve the problem because recursion still happens.

We were thinking of adding a `cache` parameter where the user can pass a dictionary. The dictionary will be filled with cached entries and it can be reused. The once one is done with generation it can just throw it away. 

An alternative is to encapsulate the generators in a class handling the cache, but it seems a bit overkill. Is there any policy about this kind of problems in NetworkX?

**Comment by dschult** [comment]
I'm not sure I understand the `cache` approach. 
We should prefer an iterative (stack) approach over recursion. But that does not inherently lead to a cache being needed.  
Each generative method can create a dict to store intermediate results -- and the dict will be deleted when the function call is done. But the user doesn't need to even know about that. When would a user need to care about the cache?

If you are thinking of implementing these methods with helper functions and the cache needs to be passed between the helper functions, then it might be better to in-line the helper functions where possible. And perhaps more importantly, it is certainly possible to make the helper functions include a `cache` input argument in order to pass the information around.

What kind of information are you cacheing?

**Comment by vigna** [comment]
The problem is related to repeated calls to the functions. Computing, for example, the number of unlabelled unrooted trees requires computing a number of convolutions, which we can do properly with dynamic programming. But if computation is repeated several times (and usually, for example for testing, one generates many trees) the computation burden of the counting functions becomes large. Deleting the dict at each call would be a gigantic waste of computational power.

An alternative approach might be having an additional parameter that specifies how many trees or forests you want to generate. Then the generating function could reuse the values computed. If you specify more than one, a list rather than a scalar could be returned.

**Reply by dschult** [reply]
ooh... I like the idea of having the input parameter specify how many trees of forests to generate. They can use the function to get this number once, store the result and use it to call the generator as many times as needed.  If I understand right that relieves the need for the large dict. :)

**Comment by vigna** [comment]
Well, it relieves the need to take the dictionary around. We will compute the counting function in linear time using linear space in all cases. But that's acceptable.

Are you OK with the idea of returning a tree or a vector of trees? This is to avoid breaking the current interface. We would have a parameter named, say, "vector" or something like that. When it's None (default) we return a tree, as now, otherwise we return a vector of the given size (or a pair of vectors, one with the trees and one with the roots, in the case of rooted forests).

Sounds OK?

**Comment by dschult** [comment]
I had an idea this morning. The trees that we return are NetworkX graphs of some sort right?  What do you think about storing the root information in the graph as a graph attribute.  Would that avoid having to return 2-tuples of (tree, root) info?

I think your idea of an input to indicate the number of trees to return. As you say, it could default to `None` with a single tree returned. That helps backward compatibility. But it could be used to create a forest of trees. I think a forest could be returns as a tuple of trees or a list of trees. As for the input parameter name, vector is a very generic term. Perhaps something like: `number_of_trees` or `forest_size`?

**Comment by vigna** [comment]
Yes, that's totally possible if you prefer it, and indeed one of the authors prefers this solution.

Well, for what function do you mean that? Because there are functions returning trees and a function returning forests. Clearly both names would be wrong for the function returning forests. You could have `number_of_trees` for the tree functions and `number_of_forests` for the forest functions.

`forest_size` is not good for trees because we will be returning a vector of unlabelled trees selected uniformly at random, but the vector of tree we return will not be a forest of unlabelled trees selected uniformly at random, as we're not considering the possible isomorphisms between the trees of the vector.

**Reply by dschult** [reply]
> Clearly both names would be wrong for the function returning forests.
:) :) Indeed!

And you people should probably be picking the names anyway -- I was just wanting something more specific than `vector`. :}

**Comment by vigna** [comment]
We're almost done. Question: how do you want the roots to be labeled? Like, a label named "root" with value `True`? And, should we label with "root" _all_ nodes except the root with `False`?

**Reply by dschult** [reply]
I was thinking you could use a graph attribute like:  `G.graph["root"] = node`. Then the root could be any node -- there would be a unique root for the graph -- and you wouldn't need to search through all the nodes to find the root. Maybe O(1) lookup of root is perhaps important(?)

**Reply by vigna** [reply]
We can do that, but then what about forests? We have a "roots" attribute listing the root of each tree in the forest?

**Reply by dschult** [reply]
That's a nice option -- then to tell if a graph is a tree or a forest you could see if there is a "root" graph attribute or a "roots" graph attribute.  ?I guess?

**Comment by vigna** [comment]
We're almost done. We met two obstacles.

The first: since by default for backward compatibility graphs are labeled, using a uniform interface our function for generating rooted forests would have thrown an exception. Fortunately a few years ago a generalization of Prüfer sequences for labeled rooted forests was discovered, so we are going to implement that.

The other problem is that random_tree has an undocumented behavior if you pass a create_using parameter creating a directed graph. In that case it visits the graph in DFS forgetting arc directions starting from node 0 and returns the out-directed tree thus obtained.

Beside the fact that the documentation states that the returned tree is represented as an undirected graph, the result is not uniformly random at all: its root is always zero, for a start.

In our code we would prefer to raise an exception if we detect that we are getting a directed graph. We would leave the bizarre current behavior just for the case of labeled rooted trees for backward compatibility, even if frankly we can't fathom in which scenario it might make sense.

**Comment by dschult** [comment]
For the second obstacle, I've been trying to figure out the trouble... In what sense does the DFS forget the edge directions? My quick look at the code makes me think it is using successors-only for the DFS. Do you have a small example?  My thought is that the directed trees should have all edges point away from the root. But I haven't looked at this for a long time.  The PR that put the directed option in place was [#4294(comment)](https://github.com/networkx/networkx/pull/4294#issuecomment-727284008)

As far as naming the root 0, it shouldn't matter what the names are right? So long as the trees are chosen uniformly the labels themselves don't matter.... or am I missing something. The same comment talks about this characteristic as well.

**Comment by vigna** [comment]
Yeah, I didn't explain myself well. I apologize.

I've read the PR: the key comment is "I did make one slight functionality change. Previously if you requested a directed graph, it would use dfs starting from an random node as the source. Now, I'm always using the node 0 as a source. It makes the function simpler, and the user can always relabel if they want random labels too.".

[I had written here a mathematical comment that was incorrect, so I removed it.]

This behavior (or the idea of relabeling randomly to obtain a rooted tree drawn uniformly at random) is not documented anywhere. And it is a completely absurd design choice: if I generate a tree with a million nodes, instead of paying the price of a single randint() call I'll have to _permute a million node graph_. I can't understand this, I'm sorry. Python is slow, OK, but this is simply too much.

From a behavioral viewpoint, if you have a method called "random_tree" that returns a _rooted_ random tree there's something broken. Because in the undirected case you're not returning a root. You're returning a different combinatorial object, depending on create_using, which was completely confusing for me, and I'm sure it will be for other users, in particular because the documentation clearly state that the function will return an undirected tree.

I mean, at least the documentation should be fixed. But I think that the documentation is absolutely reasonable. These function should not return directed structures. If the users want to turn them in to a directed structure then the user should decide how to do this, maybe with a support function. Because some users might prefer in-directed and some out-directed.

**Comment by vigna** [comment]
To be more constructive, I'd do as follows:

- These functions return undirected trees. They raise an exception if create_using gives a directed tree.
- For the only existing case, we retain the behavior but we print a warning.
- We provide in utils functions out_directed_tree and in_directed_tree that return an (in/out) directed version of a tree given a root. If the provided root is None, they try to fetch the "root" attribute of the graph. If neither are there, an exception is raised.

In this way we separate the concerns of generation and orientation.

**Comment by dschult** [comment]
I am probably mistaken in some subtlety here... but my understanding is that "uniformly random unlabeled trees" are chosen with equal weight from among the distinct (non-isomorphic) trees on n nodes. Each of these unlabeled trees has n! ways to be labeled. So, to select "uniformly random labeled trees", you can use a "uniformly random unlabeled tree" method followed by a uniformly selected relabeling on n items (a "shuffle"). Shuffles only take n random selections, so are not slow, even if there are n! possibilities. 

Turned around, if we have a way to select "uniformly random labeled trees" (that is, trees chosen with equal weight from among the distinct labeled trees -- isomorphic trees are counted separately), then we can build a uniformly random **unlabeled** tree by building the labeled tree followed by removing the labels... That is because each unlabeled tree has n! representatives in the population of labeled trees.

For rooted trees, one node from the tree is designated as the root, and isomorphism now includes the property of being the root in addition to the edge relations of graph isomorphism. So, in a similar way to the above, if we have an unlabeled method to generate uniformly random rooted trees with a root always called `0`, we can build a labeled method by adding a shuffle. And *vice versa* by stripping off the labels.

Cayley's formula says there are n^{n-2} labeled trees on n>=2 nodes. And the [paper by Aldous](https://escholarship.org/content/qt2678m3vd/qt2678m3vd_noSplash_dcf277bb94b25162c413e0b7ec1dbcc4.pdf) says (pg5) that 

> A rooted tree is merely a tree with one vertex distinguished and called the root. Isomorphisms between rooted trees are required to preserve the root. If we insist that the root be vertex 1, then there are n^{n-2} rooted labelled trees on n vertices; if not, there are n^{n-1}.

So, it looks like we can call the root 0 and then shuffle the nodes -- but I can't quite convince myself that is true...

Am I understanding the subtleties of labeled vs unlabeled correctly? For some reason I used to think that uniformly over labeled trees and uniformly over unlabeled trees followed by stripping the labels gave a different distribution of trees. But I've either figured something out, or regained my innocence and I can't quite tell which. :}

**Comment by vigna** [comment]
I'm sorry but it doesn't work like that. 

The problem is that when you label in all possible ways a graph, sometimes you get the SAME graph. Extreme case: the graph with no edges. You relabel it in n! ways and it's always the same.

If you do as you suggest, pick a graph up to isomorphism and label it in all possible ways, you will massively overrepresent graphs that have a small automorph.

The same problem happens in the opposite direction. You cannot extract a graph with random_tree and forget the labels. You will overrepresent some graph and underelresent others. That's why the algorithms in the two cases are so different.

Try this: generate twenty trees of size four with the current random_tree implementation. Look at the shapes (there are only two) and tell me if they appear the same number of times.

The problem that lead us here is that if you want to evaluate the average behavior of something on trees random_tree is useless. You need a function extracting trees uniformly at random up to isomorphism.

**Comment by vigna** [comment]
So, what about the proposal above?

**Comment by dschult** [comment]
Sorry for the delay -- I got distracted -- and that might have been a good thing because I wasn't able to focus well when I wrote those most recent comments...  so thanks for the gentle nudge. :}

Given the confusion I displayed a month ago about labeled and unlabeled random trees, I now wonder whether it would be better to have two different functions -- one for labeled and one for unlabeled.  Do you have a preference?

I think it is reasonable to have this function always return a `networkx.Graph` object.  And rather than creating separate functions for constructing a directed graph, we can include an example in the doc_string where the directed nature is provided using `bfs_tree` for example.  This does break backward compatibility, and so might impact the decision about whether to have two functions or one. We would want to deprecate the case when `create_using is not None` with a warning that it would change in version 3.4. The other cases could select a root at random and build the networkx.Graph without caring about `create_using`.

Thoughts?

**Comment by vigna** [comment]
Well, separate functions was the logic of our original PR. On Mar 14 you suggested merging and putting in a parameter. But we kept isolated the code, so rolling back is not a big issue. However, we should introduce random_labeled_tree and deprecate random_tree, otherwise the API will be confusing.

I totally support the idea of dropping create_using. random_tree can just behave as now for backward compatibility, with a warning. Doing this with separate functions is much easier because all the new functions won't simply have create_using.

**Comment by vigna** [comment]
Should we proceed?

**Comment by dschult** [comment]
Yes -- this all looks like a good plan to me.  Thanks!
